### PR TITLE
feat: [IAC-2023]: IACM plugin for IDP

### PR DESCRIPTION
**Pull request by simonharness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.

### CODE REVIEW

File: plugins/harness-iacm/src/components/ExampleFetchComponent/ExampleFetchComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+import { makeStyles } from '@material-ui/core/styles';
+import {
+  Table,
+  TableColumn,
+  Progress,
+  ResponseErrorPanel,
+} from '@backstage/core-components';
+import useAsync from 'react-use/lib/useAsync';
+
+export const exampleUsers = {
```
Comment (Debanitrkl) [inline plugins/harness-iacm/src/components/ExampleFetchComponent/ExampleFetchComponent.tsx:11]: Do we require so many example users??

File: plugins/harness-iacm/src/components/Router.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import { Routes, Route } from 'react-router';
+import WorkspaceList from './WorkspaceList';
+// import { harnessFeatureRouteRef } from '../route-refs'
```
Comment (Debanitrkl) [inline plugins/harness-iacm/src/components/Router.tsx:21]:
```suggestion
```

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+  const classes = useStyles();
+  const discoveryApi = useApi(discoveryApiRef);
+  const backendBaseUrl = discoveryApi.getBaseUrl('proxy');
+
+  const { isWorkspaceAnnotationPresent, harnessWorkspaceUrlObject } =
+    useProjectUrlSlugEntity();
+
+  const allEnvsAnnotations = [
+    ['prod', 'harness.io/project-url'],
+    ['stage', 'harness.io/project-url-stage'],
```
Comment (Debanitrkl) [inline plugins/harness-iacm/src/components/WorkspaceList/index.tsx:71]:
```suggestion
```
we no longer support stress or stage its all on devspaces now

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    ['qa', 'harness.io/project-url-qa'],
+    ['stress', 'harness.io/project-url-stress'],
```
Comment (Debanitrkl) [inline plugins/harness-iacm/src/components/WorkspaceList/index.tsx:73]:
```suggestion
```
same as above

File: plugins/harness-iacm/src/components/ExampleFetchComponent/ExampleFetchComponent.tsx (hunk 1 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/ExampleFetchComponent/ExampleFetchComponent.tsx:11]: i dont think we really need this component, I will add a proper test for the components I have added

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx (hunk 3 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/WorkspaceList/index.tsx:71]: removed, thanks

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx (hunk 4 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/WorkspaceList/index.tsx:73]: done

File: plugins/harness-iacm/src/plugin.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import { rootRouteRef } from './routes';
+
+export const harnessIacmPlugin = createPlugin({
+  id: 'harness-iacm',
+  routes: {
+    root: rootRouteRef,
+  },
+});
+
+export const HarnessIacmPage = harnessIacmPlugin.provide(
```
Comment (jenil-harness) [inline plugins/harness-iacm/src/plugin.ts:15]: Please follow this
https://backstage.io/docs/plugins/composability/#naming-patterns

Comment (jenil-harness) [inline plugins/harness-iacm/src/plugin.ts:15]: Please follow this Naming Pattern
https://backstage.io/docs/plugins/composability/#naming-patterns

File: plugins/harness-iacm/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# harness-iacm
```
Comment (jenil-harness) [inline plugins/harness-iacm/README.md:1]: Update the Docs. Without Docs I am also not able to understand what this plugin is doing.
You can refer CI/CD Plugin's docs.

File: examples/catalog-harness-iacm.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+apiVersion: backstage.io/v1alpha1
+kind: Component
+metadata:
+  name: demo-service-with-module
+  description: pipeline build and deploy
+  annotations:
+    harness.io/pipelines: |
```
Comment (jenil-harness) [inline examples/catalog-harness-iacm.yaml:7]: We can remove it from here as it is not relevant to IACM plugin

File: packages/app/src/App.tsx [hunk 8]
```diff
@@ -93,6 +94,7 @@ const routes = (
     </Route>
     <Route path="/settings" element={<UserSettingsPage />} />
     <Route path="/catalog-graph" element={<CatalogGraphPage />} />
+    <Route path="/harness-iacm" element={<EntityIacmContent />} />
```
Comment (jenil-harness) [inline packages/app/src/App.tsx:97]: We can remove this

File: packages/app/src/components/catalog/EntityPage.tsx [hunk 9]
```diff
@@ -296,6 +306,9 @@ const websiteEntityPage = (
     <EntityLayout.Route path="/ci-cd" title="CI/CD">
       {cicdContent}
     </EntityLayout.Route>
+    <EntityLayout.Route path="/ harness-iacm" title="IACM">
```
Comment (jenil-harness) [inline packages/app/src/components/catalog/EntityPage.tsx:309]:
```suggestion
    <EntityLayout.Route path="/harness-iacm" title="IACM">
```

File: plugins/harness-iacm/README.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+- Set the value of target to your on-prem URL if you are using the Harness on-prem offering
+
+3. Inside your Backstage's `EntityPage.tsx`, update the `cicdContent` component to render `<EntityHarnessCiCdContent />` whenever the service is using Harness CI/CD. Something like this -
+
+```tsx
+// In packages/app/src/components/catalog/EntityPage.tsx
+
+import {
+  isHarnessCiCdAvailable,
+  EntityHarnessCiCdContent,
```
Comment (jenil-harness) [inline plugins/harness-iacm/README.md:57]: Update.

File: plugins/harness-iacm/README.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+<details>
+  <summary>Instructions in case of separate CI and CD provider (Click to expand)</summary>
+
+```tsx
+// In packages/app/src/components/catalog/EntityPage.tsx
+
+import {
+  isHarnessCiCdAvailable,
+  EntityHarnessCiCdContent,
```
Comment (jenil-harness) [inline plugins/harness-iacm/README.md:79]: Update

File: plugins/harness-iacm/README.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+} from '@harnessio/backstage-plugin-iacm';
+
+const cicdContent = (
```
Comment (jenil-harness) [inline plugins/harness-iacm/README.md:60]: Update

File: plugins/harness-iacm/README.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+const serviceEntityPage = (
+  // ...
+  <EntityLayout.Route
+    path="/harness-ci-cd"
+    title="Harness CI/CD"
+    if={isHarnessCiCdAvailable}
+  >
+    <EntityHarnessCiCdContent />
+  </EntityLayout.Route>
+  // ...
```
Comment (jenil-harness) [inline plugins/harness-iacm/README.md:91]: update

File: plugins/harness-iacm/catalog.yaml [hunk 14]
```diff
@@ -0,0 +1,35 @@
+apiVersion: backstage.io/v1alpha1
```
Comment (jenil-harness) [inline plugins/harness-iacm/catalog.yaml:1]: We can remove this file.

File: plugins/harness-iacm/package.json [hunk 15]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@internal/backstage-plugin-harness-iacm",
```
Comment (jenil-harness) [inline plugins/harness-iacm/package.json:2]:
```suggestion
  "name": "@harnessio/backstage-plugin-harness-iacm",
```

File: plugins/harness-iacm/src/components/Router.tsx [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+import React from 'react';
+import { Routes, Route } from 'react-router';
+import WorkspaceList from './WorkspaceList';
+import { Entity } from '@backstage/catalog-model';
+import { useEntity } from '@backstage/plugin-catalog-react';
+import { MissingAnnotationEmptyState } from '@backstage/core-components';
+
+/** @public */
+export const isHarnessIacmAvailable = (entity: Entity) =>
+  Boolean(entity.metadata.annotations?.['harness.io/project-url']);
```
Comment (jenil-harness) [inline plugins/harness-iacm/src/components/Router.tsx:27]:
```suggestion
  Boolean(entity.metadata.annotations?.['harness.io/workspace-url']);
```

File: plugins/harness-iacm/src/components/Router.tsx [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+/** @public */
+
+export const Router = () => {
+  const { entity } = useEntity();
+  if (!isHarnessIacmAvailable(entity)) {
+    return (
+      <>
+        <MissingAnnotationEmptyState annotation="harness.io/project-url" />
```
Comment (jenil-harness) [inline plugins/harness-iacm/src/components/Router.tsx:36]:
```suggestion
        <MissingAnnotationEmptyState annotation="harness.io/workspace-url" />
```

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+  const [refresh, setRefresh] = useState(false);
+
+  const classes = useStyles();
+  const discoveryApi = useApi(discoveryApiRef);
+  const backendBaseUrl = discoveryApi.getBaseUrl('proxy');
+
+  const { isWorkspaceAnnotationPresent, harnessWorkspaceUrlObject } =
+    useProjectUrlSlugEntity();
+
+  const allEnvsAnnotations = [
```
Comment (jenil-harness) [inline plugins/harness-iacm/src/components/WorkspaceList/index.tsx:69]: Relook.

File: examples/catalog-harness-iacm.yaml (hunk 7 above)
Comment (simonharness) [reply examples/catalog-harness-iacm.yaml:7]: done

File: packages/app/src/App.tsx (hunk 8 above)
Comment (simonharness) [reply packages/app/src/App.tsx:97]: done

File: plugins/harness-iacm/README.md (hunk 10 above)
Comment (simonharness) [reply plugins/harness-iacm/README.md:57]: done

File: plugins/harness-iacm/README.md (hunk 12 above)
Comment (simonharness) [reply plugins/harness-iacm/README.md:60]: done

File: plugins/harness-iacm/README.md (hunk 11 above)
Comment (simonharness) [reply plugins/harness-iacm/README.md:79]: done

File: plugins/harness-iacm/README.md (hunk 13 above)
Comment (simonharness) [reply plugins/harness-iacm/README.md:91]: done

File: plugins/harness-iacm/catalog.yaml (hunk 14 above)
Comment (simonharness) [reply plugins/harness-iacm/catalog.yaml:1]: done

File: plugins/harness-iacm/package.json (hunk 15 above)
Comment (simonharness) [reply plugins/harness-iacm/package.json:2]: done

File: plugins/harness-iacm/src/components/Router.tsx (hunk 16 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/Router.tsx:27]: done

File: plugins/harness-iacm/src/components/Router.tsx (hunk 17 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/Router.tsx:36]: done

File: plugins/harness-iacm/src/components/WorkspaceList/index.tsx (hunk 18 above)
Comment (simonharness) [reply plugins/harness-iacm/src/components/WorkspaceList/index.tsx:69]: refactored

File: plugins/harness-iacm/package.json [hunk 19]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@harnessio/backstage-plugin-harness-iacm",
+  "version": "0.1.0",
+  "main": "src/index.ts",
+  "types": "src/index.ts",
+  "license": "Apache-2.0",
+  "private": true,
```
Comment (Debanitrkl) [inline plugins/harness-iacm/package.json:7]:
```suggestion
  "private": false,
```

**Review by Debanitrkl — CHANGES_REQUESTED** [review]

File: plugins/harness-iacm/package.json (hunk 19 above)
Comment (simonharness) [reply plugins/harness-iacm/package.json]: done

**Review by Debanitrkl — APPROVED** [review]
LGTM!

**Review by jenil-harness — APPROVED** [review]
LGTM

**Pull request merged**